File: services/mirror-worker/src/replay/event-splitted.ts
import axios from 'axios';
import { PoolClient } from 'pg';
import { config } from '../config';
import { log } from '../log';
import { ProdUserLog } from '../db/prod';
import { insertMapping } from '../db/staging';
import { getStagingApiToken } from '../runtime';
import { translateEquipmentEventId, translateEquipmentId } from '../translate';

// Prod payload shape (sampled session 55):
//   {
//     events: [
//       { startTime, endTime, note, type, machineCode, categoryCode,
//         subcategoryCode, descCategory, descSubcategory, idle, changeOver,
//         plannedDowntime }, ...
//     ],
//     eventType: 'downtime' | 'manual' | 'low_speed',
//     idEquipment: <prod>,
//     idEquipmentEvent: <prod>
//   }
//
// Translation: same chain as event-justified / event-edited
//   idEquipment      -> packml_topic
//   idEquipmentEvent -> interval-overlap (mirror_id_map cached)
//
// The events array passes through verbatim — operator-set metadata only,
// no IDs to translate. Prod's EventDto has a `type` field that staging's
// EventDto doesn't declare; class-validator's default behavior keeps it.
//
// Output: POST /api/downtimes/split
interface SplitPayload {
  events: unknown[];
  eventType: string;
  idEquipment: number;
  idEquipmentEvent: number;
}

export async function replayEventSplitted(
  client: PoolClient,
  row: ProdUserLog,
): Promise<void> {
  const payload = row.payload as SplitPayload;
  if (!payload || typeof payload !== 'object') {
    throw new Error('event-splitted payload missing or not an object');
  }
  if (!Array.isArray(payload.events) || payload.events.length === 0) {
    throw new Error('event-splitted payload missing non-empty events array');
  }
  if (typeof payload.idEquipment !== 'number') {
    throw new Error('event-splitted payload missing idEquipment');
  }
  if (typeof payload.idEquipmentEvent !== 'number') {
    throw new Error('event-splitted payload missing idEquipmentEvent');
  }
  if (!payload.eventType) {
    throw new Error('event-splitted payload missing eventType');
  }

  const stagingEquipmentId = await translateEquipmentId(payload.idEquipment);
  const stagingEventId = await translateEquipmentEventId(
    BigInt(payload.idEquipmentEvent),
  );
  if (stagingEventId === undefined) {
    throw new Error(
      `equipment_event ${payload.idEquipmentEvent} unmapped (no staging interval overlap; see WARN log)`,
    );
  }

  // Cache the mapping so subsequent operations on the same prod event skip
  // the SELECT — same pattern as event-justified / event-edited.
  await insertMapping(client, {
    entityType: 'equipment_event',
    source: config.sourceName,
    prodId: BigInt(payload.idEquipmentEvent),
    stagingId: stagingEventId,
    sourceLogId: BigInt(row.id_user_logs),
  });

  const body = {
    idEquipment: stagingEquipmentId,
    idEquipmentEvent: Number(stagingEventId),
    eventType: payload.eventType,
    events: payload.events,
  };

  const url = `${config.stagingApi.baseUrl}/api/downtimes/split`;
  const res = await axios.post(url, body, {
    params: {
      token: getStagingApiToken(),
      idEnterprise: config.stagingEnterpriseId,
    },
    headers: {
      'x-user': row.nm_user ?? 'mirror-worker',
      'X-Mirror-Source': config.sourceName,
    },
    validateStatus: () => true,
    timeout: 15_000,
  });

  if (res.status >= 400) {
    throw new Error(
      `staging /api/downtimes/split returned ${res.status}: ${JSON.stringify(res.data)}`,
    );
  }

  log.info(
    {
      sourceLogId: row.id_user_logs,
      prodEventId: payload.idEquipmentEvent,
      stagingEventId: stagingEventId.toString(),
      splitInto: payload.events.length,
      status: res.status,
    },
    'replayed event-splitted',
  );
}

File: services/mirror-worker-go/internal/replay/event_splitted.go
package replay

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/config"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/db"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/translate"
)

type SplitPayload struct {
	Events           []any  `json:"events"`
	EventType        string `json:"eventType"`
	IDEquipment      int    `json:"idEquipment"`
	IDEquipmentEvent int64  `json:"idEquipmentEvent"`
}

// stagingISOFormat matches JS `new Date(...).toISOString()` — split.service.ts
// runs that conversion on both sides of its equality check, so the layout we
// send needs to round-trip through it cleanly. Millisecond precision, UTC.
const stagingISOFormat = "2006-01-02T15:04:05.000Z"

// EventSplitted replays an operator split. The operator's payload carries
// THEIR timestamps — first event start = original event start, last event
// end = original event end. split.service.ts validates those equalities
// strictly against the staging row's ts_event/ts_end. Because prod runs
// CPAC 5-min smoothing while staging carries raw PLC transitions, prod's
// boundaries don't always match staging's boundaries even when the matcher
// correctly identifies the same physical event — yielding 400s like DLQ id
// 282 (prod start 2026-06-22 17:06, staging start 2026-06-22 20:41:07, same
// end 2026-06-25 06:54).
//
// Fix: rewrite events[0].startTime and events[N-1].endTime to staging's
// boundaries before sending. Operator's interior split points (the actual
// annotation transitions they care about) pass through unchanged. If staging
// is still open (ts_end IS NULL), leave the operator's endTime as-is —
// staging will close the event at that timestamp (see edge-api PR #130).
func EventSplitted(
	cfg *config.Config,
	t *translate.Translator,
	apiToken func() string,
	httpc *http.Client,
	logger *slog.Logger,
) Handler {
	return func(ctx context.Context, tx pgx.Tx, row db.ProdUserLog) error {
		var p SplitPayload
		if err := json.Unmarshal(row.Payload, &p); err != nil {
			return fmt.Errorf("parse event-splitted payload: %w", err)
		}
		if p.IDEquipment == 0 || p.IDEquipmentEvent == 0 || p.EventType == "" {
			return fmt.Errorf("event-splitted payload missing fields: %+v", p)
		}
		if len(p.Events) == 0 {
			return fmt.Errorf("event-splitted payload missing non-empty events array")
		}

		stagingEqID, err := t.Equipment(ctx, p.IDEquipment)
		if err != nil {
			return fmt.Errorf("translate equipment: %w", err)
		}
		stagingEventID, err := t.EquipmentEvent(ctx, p.IDEquipmentEvent)
		if err != nil {
			if errors.Is(err, translate.ErrUnmapped) {
				return fmt.Errorf("equipment_event %d unmapped (no staging interval overlap; see WARN log): %w",
					p.IDEquipmentEvent, err)
			}
			return fmt.Errorf("translate equipment_event: %w", err)
		}

		if err := db.InsertMapping(ctx, tx, db.MapInsert{
			EntityType:  "equipment_event",
			Source:      cfg.SourceName,
			ProdID:      p.IDEquipmentEvent,
			StagingID:   stagingEventID,
			SourceLogID: row.IDUserLogs,
		}); err != nil {
			return fmt.Errorf("insert mapping: %w", err)
		}

		stagingTsEvent, stagingTsEnd, err := t.EquipmentEventBoundaries(ctx, stagingEventID)
		if err != nil {
			return fmt.Errorf("read staging boundaries: %w", err)
		}
		startDriftSec, endDriftSec := alignSplitEventsToStaging(p.Events, stagingTsEvent, stagingTsEnd)

		body := map[string]any{
			"idEquipment":      stagingEqID,
			"idEquipmentEvent": stagingEventID,
			"eventType":        p.EventType,
			"events":           p.Events,
		}
		status, _, err := PostStaging(ctx, cfg, httpc, apiToken(), row,
			"/api/downtimes/split", body)
		if err != nil {
			return err
		}
		logger.Info("replayed event-splitted",
			slog.Int64("sourceLogID", row.IDUserLogs),
			slog.Int64("prodEventID", p.IDEquipmentEvent),
			slog.Int64("stagingEventID", stagingEventID),
			slog.Int("splitInto", len(p.Events)),
			slog.Int("startDriftSec", startDriftSec),
			slog.Int("endDriftSec", endDriftSec),
			slog.Int("status", status))
		return nil
	}
}

// alignSplitEventsToStaging mutates events[0].startTime to stagingTsEvent
// and (when stagingTsEnd is non-null) events[N-1].endTime to stagingTsEnd.
// Returns the drift in seconds for log diagnostics — large drift indicates
// the matcher chose a staging event with significantly different boundaries
// than prod's, which is informational but not actionable (the match is
// valid; the boundaries simply differ).
//
// Type assertions are best-effort: if events[i] isn't a JSON object after
// decode (shouldn't happen for real payloads but might for malformed ones),
// we skip the rewrite for that element and let staging reject the original.
func alignSplitEventsToStaging(events []any, stagingTsEvent time.Time, stagingTsEnd sql.NullTime) (startDriftSec, endDriftSec int) {
	first, ok := events[0].(map[string]any)
	if !ok {
		return 0, 0
	}
	if prev, ok := first["startTime"].(string); ok {
		if t, err := time.Parse(time.RFC3339Nano, prev); err == nil {
			startDriftSec = int(stagingTsEvent.Sub(t).Seconds())
		}
	}
	first["startTime"] = stagingTsEvent.UTC().Format(stagingISOFormat)

	if !stagingTsEnd.Valid {
		return startDriftSec, 0
	}
	last, ok := events[len(events)-1].(map[string]any)
	if !ok {
		return startDriftSec, 0
	}
	if prev, ok := last["endTime"].(string); ok {
		if t, err := time.Parse(time.RFC3339Nano, prev); err == nil {
			endDriftSec = int(stagingTsEnd.Time.Sub(t).Seconds())
		}
	}
	last["endTime"] = stagingTsEnd.Time.UTC().Format(stagingISOFormat)
	return startDriftSec, endDriftSec
}
